lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/CachingHdfsInputFile.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.hdfs;
+
+import alluxio.client.file.CacheContext;
+import alluxio.client.file.URIStatus;
+import alluxio.client.file.cache.CacheManager;
+import alluxio.client.file.cache.LocalCacheFileInStream;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.hadoop.AlluxioHdfsInputStream;
+import alluxio.hadoop.HdfsFileInputStream;
+import alluxio.wire.FileInfo;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hashing;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoInput;
+import io.trino.hdfs.CallStats;
+import io.trino.hdfs.HdfsContext;
+import io.trino.hdfs.HdfsEnvironment;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.io.IOException;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class CachingHdfsInputFile
+        extends HdfsInputFile
+{
+    private final CacheManager cacheManager;
+    private final AlluxioConfiguration alluxioConf;
+    private final FileSystem.Statistics statistics = new FileSystem.Statistics("alluxio");
+    private final HashFunction hashFunction = Hashing.murmur3_128();
+
+    public CachingHdfsInputFile(Location location, Long length, Long lastModifiedTime, HdfsEnvironment environment, HdfsContext context, CallStats callStat,
+                                CacheManager cacheManager, AlluxioConfiguration alluxioConf)
+    {
+        super(location, length, lastModifiedTime, environment, context, callStat);
+        this.cacheManager = cacheManager;
+        this.alluxioConf = alluxioConf;
+    }
+
+    @Override
+    public TrinoInput newInput()
+            throws IOException
+    {
+        FileSystem fileSystem = environment.getFileSystem(context, file);
+        FSDataInputStream input = environment.doAs(context.getIdentity(), () -> {
+            String path = file.toString();
+            long lastModifiedTime = lastModified().toEpochMilli();
+
+            FileInfo info = new FileInfo()
+                    .setLastModificationTimeMs(lastModifiedTime)
+                    .setPath(path)
+                    .setFolder(false)
+                    .setLength(length());
+            String cacheIdentifier = hashFunction.hashString(path + lastModifiedTime, UTF_8).toString();
```

Comment:
It's a bit unfortunate that we have to do the caching on the file level, and not on the part-of-file level. With the current caching scheme a single large file from cloud storage will be cached on a single node. For a large and hot file this could potentially lead to problems. But it might be hard to solve, and might not be worth it.

Reply:
@beinan could you please weigh in on whether this is possible to improve ?